File: 2010/J101.pas
const
  // 10 = 0 left, 11 = 0 right
  dist:array[1..11] of string[12] = (
    '01212323412',
    '10121232321',
    '21032143232',
    '12301212323',
    '21210121232',
    '32121032143',
    '23412301234',
    '32321210143',
    '43232121054',
    '12323434501',
    '21232343410');
var
  s: ansistring;
  location, moves, i, len: longint;
begin
  read(s);
  moves := 0;
  location := 5;
  len := length(s);
  for i := 1 to len do
  begin
    if ((s[i] = '0') and (location <> 10) and (location <> 1) and (location <> 4) and (location <> 7)) then
    begin
      moves := moves + ord(dist[location][11]) - ord('0');
      location := 11
    end
    else if (s[i] = '0') then
    begin
      moves := moves + ord(dist[location][10]) - ord('0');
      location := 10
    end
    else
    begin
      moves := moves + ord(dist[location][ord(s[i]) - ord('0')]) - ord('0');
      location := ord(s[i]) - ord('0')
    end
  end;
  writeln(moves)
end.

File: 2010/J101.cpp
#include <cstdio>
#include <cstring>
int main() {
  // 0 = 0 left, 10 = 0 right
  char dist[11][12] = {"01232343451",
                       "10121232342",
                       "21012123231",
                       "32103214322",
                       "21230121233",
                       "32121012122",
                       "43212103213",
                       "32341230124",
                       "43232121013",
                       "54323212104",
                       "12123234340"};
  char s[32];
  scanf("%s", s);
  int location = 5, moves = 0;
  int len = strlen(s);
  for (int i = 0; i < len; i++) {
    if (s[i] == '0' && location != 0 && location != 1 && location != 4 && location != 7) {
      moves += dist[location][10] - '0';
      location = 10;
    } else if (s[i] == '0') {
      moves += dist[location][0] - '0';
      location = 0;
    } else {
      moves += dist[location][s[i] - '0'] - '0';
      location = s[i] - '0';
    }
  }
  printf("%d\n", moves);
  return 0;
}
